### ISSUE TITLE
Remove `std::function<>` from the `Ledger::AttestPromotion()` flow

### ISSUE THREAD

**Issue by szilardszaloki** [brave/brave-browser#24294]
Part4 of https://github.com/brave/brave-browser/issues/18560.

**Comment by GeetaSarvadnya** [comment]
Verification is inprogress

```

Brave | 1.43.67 Chromium: 104.0.5112.81 (Official Build) beta (64-bit)
-- | --
Revision | 5b7b76419d50f583022568b6764b630f6ddc9208-refs/branch-heads/5112@{#1309}
OS | Windows 10 Version 21H2 (Build 19044.1889)

```

<details>
<summary>Unverified (no custodian connected)_claim promotion</summary>

</details>

<details>
<summary>Uphold connected_claim promotion</summary>

</details>

<details>
<summary>Gemini connected_claim promotion</summary>

</details>

<details>
<summary>bitFlyer connected_claim promotion</summary>

</details>

**Comment by LaurenWags** [comment]
Removed `QA Pass-Win64` as this is in 1.44.x and looks like it was tested with 1.43.x, so needs to be rechecked with 1.44.x.

**Comment by LaurenWags** [comment]
Verification in progress with

```
Brave | 1.44.79 Chromium: 105.0.5195.102 (Official Build) beta (x86_64)
-- | --
Revision | 4c16f5ffcc2da70ee2600d5db77bed423ac03a5a-refs/branch-heads/5195_55@{#4}
OS | macOS Version 12.5.1 (Build 21G83)
```

**Comment by GeetaSarvadnya** [comment]
Verification is inprogress  `Vivo x70 pro with android version 12` running `Bravemonoarmx64.apk_1.44.97`

**Comment by Uni-verse** [comment]
in Progress on Android ARM using the following build(s):

Brave	1.44.97 Chromium: 106.0.5249.55 (Official Build) (64-bit) 
Revision	4d5f098fca6ab7f4b6b7c240be3d9593c2357709-refs/branch-heads/5249@{#531}
OS	Android 12; Build/SP1A.210812.016

**Comment by MadhaviSeelam** [comment]
Verification  `In progress` using 
```
Brave | 1.44.97 Chromium: 106.0.5249.55 (Official Build) (64-bit)
-- | --
Revision | 4d5f098fca6ab7f4b6b7c240be3d9593c2357709-refs/branch-heads/5249@{#531}
OS | Linux
```

<details>
<summary>Unverified (no custodian connected)_claim promotion_PASSED</summary>
<details>
<summary>Case 1_Claim promotion via BR panel_PASSED</summary>

- Confirmed ad grants promotion can be claimed via rewards settings page
- Confirmed claimed grants are shown in the wallet balance
- Confirmed claimed ad grants promotion is shown in the `rewards summary`, `monthly statement` and  `brave://rewards-internal` pages

Example | Example | Example | Example | Example
-- | -- | -- | -- | --
<image width ="1200" alt="image1" src="https://user-images.githubusercontent.com/98358127/192371701-6acd2c5c-ed94-4e2b-87f7-fa11f9e7fffc.png">|<image width ="1200" alt="image1" src="https://user-images.githubusercontent.com/98358127/192371792-e781326a-19cd-401d-bf12-b895a14816f7.png">|<image width ="1200" alt="image1" src="https://user-images.githubusercontent.com/98358127/192372049-0c89b651-776f-447a-900a-45652a1c68e2.png">|<image width ="1200" alt="image1" src="https://user-images.githubusercontent.com/98358127/192373486-18e4d492-0579-44ff-aa88-2f92f8044061.png">|<image width ="1200" alt="image1" src="https://user-images.githubusercontent.com/98358127/192372586-3b680e2f-d91e-4e98-bed0-4cec12228fed.png">

console logs
```
[ REQUEST ]
> URL: https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=2bd7fe73-cb66-438d-a7ce-b7c924b2b562&platform=linux
> Method: GET
[15435:15435:0926/102859.402285:VERBOSE1:uphold.cc(115)] Uphold wallet is not VERIFIED.
[15435:15435:0926/102859.458253:VERBOSE6:logging_util.cc(137)] 
[ RESPONSE - OnRequest ]
> Url: https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=2bd7fe73-cb66-438d-a7ce-b7c924b2b562&platform=linux
> Result: Success
> HTTP Code: 200
> Body: {"promotions":[{"id":"fef29d18-2115-41d2-9cd3-0f43c3440aca","createdAt":"2022-09-26T16:15:09.338538Z","expiresAt":"2023-01-26T16:15:09.338538Z","version":5,"suggestionsPerGrant":44,"approximateValue":"11","type":"ads","available":true,"platform":"desktop","publicKeys":["VNjakN83bBkJkERwPDquo6QxIWBtEvNLIQKcW8SU30Y="],"legacyClaimed":false,"claimableUntil":"2022-12-26T16:15:09.338538Z"}]}
```
</details>

<details>
<summary>Case 2 - Claim promotion via settings page - PASSED</summary>


- Confirmed ad grants promotion can be claimed via rewards settings page
- Confirmed claimed grants are shown in the wallet balance
- Confirmed claimed ad grants promotion is shown in the `rewards summary`, `BR panel`, `NTP`, `monthly statement` and  `brave://rewards-internal` pages
- Confirmed claimed Ad grants moved to the `Gemini wallet`
- Confirmed claimed ad grants reflected in the `Gemini wallet balance` and `transaction history`

Example | Example | Example | Example | Example 
-- | -- | -- | -- | -- 
<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192374363-25efd21c-b3b5-456c-8ad6-1f508856e403.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192374504-6b1d42af-d699-4fdd-92a0-ac7c2237bb2f.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192375030-01cf8a63-fd1a-4add-8421-10bf628850ce.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192375189-052b21ac-b295-4001-8336-78b92e396432.png">|<image width="1500" alt="image1" src=" https://user-images.githubusercontent.com/98358127/192375287-de8a27bc-abfb-44a7-866e-9df622bee531.png">

console logs
```
[ REQUEST ]
> URL: https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=9716f483-d295-45ad-a3cf-377fa6b4e3a4&platform=linux
> Method: GET
[15435:15435:0926/102929.434815:VERBOSE1:uphold.cc(115)] Uphold wallet is not VERIFIED.
[15435:15435:0926/102929.490603:VERBOSE6:logging_util.cc(137)] 
[ RESPONSE - OnRequest ]
Url: https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=9716f483-d295-45ad-a3cf-377fa6b4e3a4&platform=linux
> Result: Success
> HTTP Code: 200
> Body: {"promotions":[{"id":"f05568e6-f4a1-4438-8e81-cc0490fd14f7","createdAt":"2022-09-22T14:39:18.189457Z","expiresAt":"2023-01-22T14:39:18.189457Z","version":5,"suggestionsPerGrant":120,"approximateValue":"30","type":"ugp","available":true,"platform":"desktop","publicKeys":["6AphTvx13IgxVRG1nljV2ql1Y7yGUol6yrVMhEP85wI="],"legacyClaimed":false,"claimableUntil":"2022-12-22T14:39:18.189457Z"},{"id":"1505ee92-a4de-418a-a2d7-d41ef70e1154","createdAt":"2022-09-26T19:14:19.282539Z","expiresAt":"2023-01-26T19:14:19.282539Z","version":5,"suggestionsPerGrant":32,"approximateValue":"8","type":"ads","available":true,"platform":"desktop","publicKeys":["yvj+jlP/OLP+WMBsp3iPFmeHhnCcxMzpl4dMxmo2004="],"legacyClaimed":false,"claimableUntil":"2022-12-26T19:14:19.282539Z"}]}
```
</details>

</details>

<details>
<summary>Uphold connected_claim promotion_PASSED</summary>

<details>
<summary>Case 1_Claim promotion via BR panel_PASSED</summary>

- Confirmed ad grants promotion can be claimed via rewards settings page
- Confirmed claimed grants are shown in the wallet balance
- Confirmed claimed ad grants promotion is shown in the `rewards summary`, `BR panel`, `NTP`, `monthly statement` and  `brave://rewards-internal` pages
- Confirmed claimed Ad grants moved to the `uphold wallet`
- Confirmed claimed ad grants reflected in the `uphold wallet balance` and the `transaction history`

Example | Example | Example | Example | Example 
-- | -- | -- | -- | -- 
<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192378532-8dd7652f-6447-4ae9-af06-db74deb742da.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192378875-00e7f03e-7042-4d5e-91cc-f94c2844bf1b.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192379021-8dccfa8b-5280-4d6b-bb09-f41c2cd807d2.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192379687-5645ddad-a8f6-4670-aa1a-0d5192b94d44.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192383369-8722140f-b29f-4fff-8a12-a1d610438338.png">


</details>

<details>
<summary>Case 2_Claim promotion via rewards settings page_PASSED</summary>

- Confirmed ad grants promotion can be claimed via rewards settings page
- Confirmed claimed grants are shown in the wallet balance
- Confirmed claimed ad grants promotion is shown in the `rewards summary`, `BR panel`, `NTP`, `monthly statement` and  `brave://rewards-internal` pages
- Confirmed claimed Ad grants moved to the `uphold wallet`
- Confirmed claimed ad grants reflected in the `uphold wallet balance` and the `transaction history`

Example | Example | Example | Example | Example
-- | -- | -- | -- | -- 
<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192385479-66592ece-072e-414e-8753-b132c2fe3c42.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192386467-dbefc089-cbde-491e-af85-06a9a036486d.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192386065-6ff9ce2a-b487-4e0b-836a-b27c28fb068d.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192388807-aaa2c8bb-d521-4113-bfbb-b96db0a45fb4.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192388317-2a8d1895-39d8-4d76-a922-757584fdc13c.png">


</details>
</details>

<details>
<summary>bitFlyer connected_claim promotion_PASSED</summary>

<details>
<summary>Case 1_Claim promotion via BR panel_PASSED</summary>


- Confirmed ad grants promotion can be claimed via BR panel
- Confirmed claimed grants are shown in the wallet balance
- Confirmed claimed ad grants promotion is shown in the `rewards summary`, `BR panel`, `NTP`, `monthly statement` and  `brave://rewards-internal` pages
- Confirmed claimed Ad grants moved to the `Gemini wallet`
- Confirmed claimed ad grants reflected in the `Gemini wallet balance` and `transaction history`

Example | Example | Example | Example | Example 
-- | -- | -- | -- | -- 
<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192398104-b956cac7-2e89-4667-b8bf-081e945cc819.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192398352-b1e930e2-b31a-4d3e-9cf4-5d8f619aca9f.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192398429-c561579d-f47e-4fe5-bc9e-f171a9b1b340.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192398776-d2dec409-3b45-4839-8ee6-72740870bfeb.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192398879-636d0d19-7acc-4096-a90d-a92273769953.png">

</details>
<details>
<summary>Case 2_Claim promotion via settings page_PASSED</summary>

- Confirmed ad grants promotion can be claimed via rewards settings page
- Confirmed calimed grants are shown in the wallet balance
- Confirmed claimed ad grants promotion is shown in the `rewards summary`, `BR panel`, `NTP`, `monthly statement` and  `brave://rewards-internal` pages
- Confirmed claimed Ad grants moved to the `bitFlyer wallet`
- Confirmed claimed ad grants reflected in the `bitFlyer wallet balance` 

Example | Example | Example | Example 
-- | -- | -- | -- 
<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192394647-791c9143-86eb-48a0-86e5-9d5d6842407d.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192394731-9f996485-9c57-4db9-a861-296f0a931f56.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192397000-e720fc72-740f-48a1-84af-ee63fce42a0c.png">|<image width="1500" alt="image1" src="https://user-images.githubusercontent.com/98358127/192397165-6f999eba-eaf4-4d3c-a6b8-f5d646b5714b.png">




console logs:
```
[ REQUEST ]
> URL: https://grant.rewards.bravesoftware.com/v2/suggestions/claim
> Method: POST
> Content: {"credentials":[{"publicKey":"TMAx05i2eCYqJ+R2ru06AP..."}
> Content Type: application/json; charset=utf-8
> Header digest: SHA-256=z4MCd5V4jjXN/za60ZAJ5NsA1R2jvMej+nMj25ALRDM=
> Header signature: keyId="28402e90-9a36-40f9-b7f9-d4614e1b5793",algorithm="ed25519",headers="digest (request-target)",signature="PcqNbRQljgmvYn2EgHo4+rGQdHnfSvJ7JNeMpRDzRbQtt3cVV1j8YGVOCCAqwBYMjWGcBiKYWhhHHGK+cUgJDg=="
> Header accept: application/json
[42280:42280:0926/155446.887201:VERBOSE6:logging_util.cc(137)] 
[ RESPONSE - OnRequest ]
> Url: https://demo24kiuw4dcyncsy3qlud8u8.azurewebsites.net/api/link/v1/account/inventory
> Result: Success
> HTTP Code: 200
> Body: {
  "account_hash": "f50b018b5f2b75a00c30eb286a22baf11c88cec1",
  "inventory": [
    {
      "currency_code": "JPY",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "BTC",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "BCH",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "ETH",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "ETC",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "LTC",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "MONA",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "LSK",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "XRP",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "BAT",
      "amount": 105.0,
      "available": 105.000000000000
    },
    {
      "currency_code": "XLM",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "XEM",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "XTZ",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "DOT",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "LINK",
      "amount": 0.0,
      "available": 0.0
    }
  ]
}
[42280:42280:0926/155447.032361:VERBOSE9:logging_util.cc(138)] 
[ RESPONSE HEADERS ]
> Url: https://demo24kiuw4dcyncsy3qlud8u8.azurewebsites.net/api/link/v1/account/inventory

[ REQUEST ]
> URL: https://demo24kiuw4dcyncsy3qlud8u8.azurewebsites.net/api/link/v1/account/inventory
> Method: GET
[42280:42280:0926/160246.931583:VERBOSE6:logging_util.cc(137)] 
[ RESPONSE - OnRequest ]
> Url: https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=28402e90-9a36-40f9-b7f9-d4614e1b5793&platform=linux
> Result: Success
> HTTP Code: 200
> Body: {"promotions":[{]}
[42280:42280:0926/160246.932208:VERBOSE9:logging_util.cc(138)] 
[ RESPONSE HEADERS ]
> Url: https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=28402e90-9a36-40f9-b7f9-d4614e1b5793&platform=linux

> Header content-length: 391
> Header content-type: application/json
> Header date: Mon, 26 Sep 2022 23:02:46 GMT
> Header request-id: ccp2v5kb1qnl8tn9g4q0
> Header via: 1.1 7514e5e25722778fd4b1744d4ecc67e0.cloudfront.net (CloudFront)
> Header x-amz-cf-id: FBuBSseg0qsuFzAgMVBTBfV2f8px4l1sXNssXcRmMa0-O2vxPPT1EA==
> Header x-amz-cf-pop: SEA19-C1
> Header x-cache: Miss from cloudfront
> Header x-request-id: AWkGCxKiTLf5qwYR
[42280:42280:0926/160247.033818:VERBOSE6:logging_util.cc(137)] 
[ RESPONSE - OnRequest ]
> Url: https://demo24kiuw4dcyncsy3qlud8u8.azurewebsites.net/api/link/v1/account/inventory
> Result: Success
> HTTP Code: 200
> Body: {
  "account_hash": "f50b018b5f2b75a00c30eb286a22baf11c88cec1",
  "inventory": [
    {
      "currency_code": "JPY",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "BTC",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "BCH",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "ETH",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "ETC",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "LTC",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "MONA",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "LSK",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "XRP",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "BAT",
      "amount": 113.0,
      "available": 113.000000000000
    },
    {
      "currency_code": "XLM",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "XEM",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "XTZ",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "DOT",
      "amount": 0.0,
      "available": 0.0
    },
    {
      "currency_code": "LINK",
      "amount": 0.0,
      "available": 0.0
    }
  ]
}
[42280:42280:0926/160247.038565:VERBOSE9:logging_util.cc(138)] 
[ RESPONSE HEADERS ]
> Url: https://demo24kiuw4dcyncsy3qlud8u8.azurewebsites.net/api/link/v1/account/inventory
```
</details>
